components/script/dom/bluetooth.rs

```diff
@@ -40,12 +56,155 @@ impl Bluetooth {
     }
 }
 
+impl Clone for StringOrUnsignedLong {
+    fn clone(&self) -> StringOrUnsignedLong {
+        match self {
+            &StringOrUnsignedLong::String(ref s) => StringOrUnsignedLong::String(s.clone()),
+            &StringOrUnsignedLong::UnsignedLong(usl) => StringOrUnsignedLong::UnsignedLong(usl),
+        }
+    }
+}
+
+impl Clone for BluetoothScanFilter {
+    fn clone(&self) -> BluetoothScanFilter {
+        match self.services {
+            None => BluetoothScanFilter {
+                name: self.name.clone(),
+                namePrefix: self.namePrefix.clone(),
+                services: None,
+            },
+            Some(ref vec) => BluetoothScanFilter {
+                name: self.name.clone(),
+                namePrefix: self.namePrefix.clone(),
+                services: Some(vec.clone()),
+            },
+        }
+    }
+}
+
+impl Clone for RequestDeviceOptions {
+    fn clone(&self) -> RequestDeviceOptions {
+        match self.optionalServices {
+            None => RequestDeviceOptions {
+                filter: self.filter.clone(),
+                optionalServices: None,
+            },
+            Some(ref vec) => RequestDeviceOptions {
+                filter: self.filter.clone(),
+                optionalServices: Some(vec.clone()),
+            }
+        }
+    }
+}
+
+fn canonicalize_filter(filter: &BluetoothScanFilter, global: GlobalRef)
+    -> Fallible<BluetoothScanfilter> {
+        if !(filter.services.is_some() ||
+              filter.name.is_some() ||
+              filter.namePrefix.is_some()) {
+            return Err(Type("A filter must restrict the devices in some way.".to_owned()));
+        }
+        let filter = filter.clone();
+        let mut services: Vec<String> = vec!();
+        if filter.services.is_some() {
+            if filter.services.clone().unwrap().is_empty() {
+                return Err(Type("'services', if present, must contain at least one service.".to_owned()));
+            }
+            for service in filter.services.clone().unwrap() {
+                let validated_service = BluetoothUUID::GetService(global, service);
+                match validated_service {
+                    Err(err) => return Err(err),
+                    Ok(valid) => services.push(valid.to_string()),
+                }
+            }
+        }
+
+        let mut name = String::new();
+        if filter.name.is_some() {
+            // DOMString::len() gives back the size in bytes
+            if filter.name.clone().unwrap().len() > MAX_DEVICE_NAME_LENGTH {
```

Comment:
if I remove the clone the build gives the following error:
https://gist.github.com/zakorgy/e0d2a5e90a9f9cd73c4342d059184aa3
maybe it occurs, because this is a DOMString